drivers: usb_dc_sam0: add status interrupts

sslupsky opened the pull request:
The interrupts for SUSPEND, End of Resume and Wakeup are not enabled.

This commit enables the interrupts and executes the respective status callbacks.

sslupsky committed:
drivers: usb_dc_sam0: add status interrupts

The interrupts for SUSPEND, End of Resume and Wakeup
are not enabled.

This commit enables the interrupts and executes the respective
status callbacks.

Signed-off-by: Steven Slupsky <sslupsky@gmail.com>

jfischer-phytec-iot commented on drivers/usb/device/usb_dc_sam0.c:
```diff
@@ -254,6 +270,8 @@ int usb_dc_attach(void)
 	regs->CTRLA.bit.ENABLE = 1;
 	usb_sam0_wait_syncbusy();
 	regs->CTRLB.bit.DETACH = 0;
+	/* FIXME:  Should the state be changed to CONNECTED here */
```
Actually description of this even is weird, to be connected and "enumeratiaon is completed" are different states. Use it or remove this comments (I suggest to remove).

sslupsky replied:
Thanks for the feedback.  I intended the comment to get solicit feedback since I was uncertain if the state change should be included here.  I will remove it.

jfischer-phytec-iot replied:
:arrow_up: still not addressed

sslupsky replied:
I'll deal with this when I get the other issue resolved and do them both at the same time.

sslupsky commented on drivers/usb/device/usb_dc_sam0.c:
```diff
@@ -234,7 +247,10 @@ int usb_dc_attach(void)
+	regs->INTENSET.reg = USB_DEVICE_INTENSET_EORST |
+			     USB_DEVICE_INTENSET_EORSM |
+			     USB_DEVICE_INTENSET_SUSPEND |
+			     USB_DEVICE_INTENSET_WAKEUP;
```
Is this interrupt used for Remote Wakeup up as described in the datasheet 32.6.2.14?

jfischer-phytec-iot replied:
We probably have different datasheets. TBH I do not know how a device could use this WAKEUP interrupt in opposite to EORSM and SUSPEND. Remote Wakeup feature is to wake-up the host (29.6.3.7 Suspend and Wakeup, SAM R21)

sslupsky replied:
Apologies, I should have indicated I was referring to the SAMD21 datasheet.  I think the equivalent section in the SAMR21 datasheet is 29.6.2.14 Remote Wakeup.

My understanding is the device can initiate a remote wakeup of the host?

jfischer-phytec-iot replied:
Yes, like HID keyboard can wake-up a PC. We actually have API for that, usb_dc_wakeup_request(), see also usb_dc_nrfx driver.

sslupsky replied:
So, it is not clear to me, this is ok then?

jfischer-phytec-iot replied:
No

sslupsky replied:
So what is the issue?

jfischer-phytec-iot replied:
>You do not need to enable WAKEUP in the device mode, please remove it here and above L118..L122

sslupsky replied:
Why?  From the samd21 datasheet:

> When the controller sends the Upstream Resume INTFLAG.WAKEUP is set and INTFLAG.SUSPEND is cleared.
The CTRLB.UPRSM is cleared at the end of the transmitting Upstream Resume.

It appears that this indication has a useful purpose?

sslupsky replied:
Since I do not use the WAKEUP interrupt notification I will go ahead and remove it.  I put it in there because it appears to have some useful purpose.  So, I do think you should consider that it should be supported in the future.

jfischer-phytec-iot replied:
As I already annotated [here](https://github.com/zephyrproject-rtos/zephyr/pull/24402#discussion_r409687415), this is something for the host mode, see "Host Operations". It probably can also be used in the "Upstream Resume" procedure, but is is not the way you wanted to use it. Please remove WAKEUP comment from the commit message.

sslupsky commented:
@carlescufi @nzmichaelh What needs to happen with this PR?

nzmichaelh commented:
> @carlescufi @nzmichaelh What needs to happen with this PR?

Hi @sslupsky not sure myself.  The change looks sane, but I don't know this API well enough. Looks like @jfischer-phytec-iot has been doing most of the review.  Johann, WDYT?

Is there a test suite or compatibility suite you can run to confirm?

jfischer-phytec-iot commented:
> What needs to happen with this PR?

requested changes should be addressed.

sslupsky commented:
@jfischer-phytec-iot I responded to your comments above.  I had not heard further from you on those matters so thought my response addressed the issues.  Your recent comment suggests otherwise?

stephanosio commented on drivers/usb/device/usb_dc_sam0.c:
```diff
@@ -234,7 +242,9 @@ int usb_dc_attach(void)
 	regs->DESCADD.reg = (uintptr_t)&data->descriptors[0];
 
-	regs->INTENSET.reg = USB_DEVICE_INTENSET_EORST;
+	regs->INTENSET.reg = USB_DEVICE_INTENSET_EORST |
+			     USB_DEVICE_INTENSET_EORSM |
+			     USB_DEVICE_INTENSET_SUSPEND;
```
FYI, this is not how ASF handles EORSM and SUSPEND (these interrupts are enabled on-demand).

See #26657

Also ASF implementation:
https://github.com/adafruit/asf4/blob/d270f79aa16dd8fd4ae3b6c14544283dcb992e9c/samd51/hpl/usb/hpl_usb.c#L982-L984

https://github.com/adafruit/asf4/blob/d270f79aa16dd8fd4ae3b6c14544283dcb992e9c/samd51/hpl/usb/hpl_usb.c#L992-L994

stephanosio commented on drivers/usb/device/usb_dc_sam0.c:
```diff
@@ -262,22 +272,25 @@ int usb_dc_attach(void)
 
 	regs->CTRLA.bit.SWRST = 1;
 	usb_sam0_wait_syncbusy();
+	data->cb(USB_DC_RESET, NULL);
```
Is this necessary? Other platform drivers do not seem to do this.

Also EORST interrupt should be triggered when reset is done anyway.

stephanosio commented on drivers/usb/device/usb_dc_sam0.c:
```diff
@@ -262,22 +272,25 @@ int usb_dc_attach(void)
 	regs->CTRLB.bit.DETACH = 1;
 	usb_sam0_wait_syncbusy();
-
+	data->cb(USB_DC_DISCONNECTED, NULL);
```
Same with this. This should eventually lead to `SUSPEND` interrupt getting triggered.

`USB_DC_SUSPEND` seems to do all the necessary clean-ups and AFAICS there is no way to truly detect "detach"/"disconnect" event in the SAM0 USB device core, so `USB_DC_DISCONNECTED` should be left unimplemented (IIUC, this is optional).

stephanosio commented on drivers/usb/device/usb_dc_sam0.c:
```diff
@@ -234,7 +242,9 @@ int usb_dc_attach(void)
 
-	regs->INTENSET.reg = USB_DEVICE_INTENSET_EORST;
+	regs->INTENSET.reg = USB_DEVICE_INTENSET_EORST |
+			     USB_DEVICE_INTENSET_EORSM |
```
Also for normal "SUSPEND," we should be handling WAKEUP, not EORSM. That is exactly what ASF does too.

One caveat is that all disabled interrupts seem to be automatically "handled" (acknowledged), so leaving WAKEUP disabled does not have any lasting effect (so it seems).

stephanosio reviewed (changes requested):
We do not have any detailed public information on this peripheral beyond what is briefly described in the datasheet.

In that case, we should try to replicate whatever ASF does.

carlescufi commented:
@sslupsky you need to address the review comments or argue why they shouldn't be addressed.

sslupsky commented:
@carlescufi I fixed the commit message request that was outstanding.  Did I miss something else?

I think I addressed the outstanding comments?

carlescufi commented:
> @carlescufi I fixed the commit message request that was outstanding. Did I miss something else?
> 
> I think I addressed the outstanding comments?

You didn't reply to the comments posted so that led me to assume you hadn't, sorry. Let me re-request a review from @jfischer-phytec-iot and @stephanosio

stephanosio reviewed (changes requested):
Didn't receive any response to the previous comments.

While looking into #26657, I found that not implementing the ASF approach (enabling/disabling interrupts on the fly) leads to a lot of spurious suspend interrupts. In general, we should try to follow the ASF implementation unless we can be sure doing otherwise will not be a problem, as I noted in https://github.com/zephyrproject-rtos/zephyr/pull/26657#issuecomment-654644352.

Also, before we can claim that the suspend-resume process works on SAM0, #26725 will need to be resolved first (note that this issue is not specific to SAM0, but applies to all platforms).
